debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerQueryResultProcessor.java

```diff
@@ -0,0 +1,245 @@
+import io.debezium.connector.oracle.logminer.valueholder.LogMinerDmlEntry;
+import io.debezium.data.Envelope;
+import io.debezium.pipeline.EventDispatcher;
+import io.debezium.pipeline.source.spi.ChangeEventSource;
+import io.debezium.relational.Table;
+import io.debezium.relational.TableId;
+import io.debezium.util.Clock;
+
+/**
+ * This class process entries obtained from LogMiner view.
+ * It parses each entry.
+ * On each DML it registers a callback in TransactionalBuffer.
+ * On rollback it removes registered entries from TransactionalBuffer.
+ * On commit it executes all registered callbacks, which dispatch ChangeRecords.
+ * This also calculates metrics
+ */
+class LogMinerQueryResultProcessor {
+
+    private final ChangeEventSource.ChangeEventSourceContext context;
+    private final LogMinerMetrics metrics;
+    private final TransactionalBuffer transactionalBuffer;
+    private final SimpleDmlParser dmlParser;
+    private final OracleOffsetContext offsetContext;
+    private final OracleDatabaseSchema schema;
+    private final EventDispatcher<TableId> dispatcher;
+    private final TransactionalBufferMetrics transactionalBufferMetrics;
+    private final String catalogName;
+    private final Clock clock;
+    private final Logger LOGGER = LoggerFactory.getLogger(LogMinerQueryResultProcessor.class);
+    private long currentOffsetScn = 0;
+    private long currentOffsetCommitScn = 0;
+    private long stuckScnCounter = 0;
+
+    LogMinerQueryResultProcessor(ChangeEventSource.ChangeEventSourceContext context, LogMinerMetrics metrics,
+                                 TransactionalBuffer transactionalBuffer, SimpleDmlParser dmlParser,
+                                 OracleOffsetContext offsetContext, OracleDatabaseSchema schema,
+                                 EventDispatcher<TableId> dispatcher, TransactionalBufferMetrics transactionalBufferMetrics,
+                                 String catalogName, Clock clock) {
+        this.context = context;
+        this.metrics = metrics;
+        this.transactionalBuffer = transactionalBuffer;
+        this.dmlParser = dmlParser;
+        this.offsetContext = offsetContext;
+        this.schema = schema;
+        this.dispatcher = dispatcher;
+        this.transactionalBufferMetrics = transactionalBufferMetrics;
+        this.catalogName = catalogName;
+        this.clock = clock;
+    }
+
+    /**
+     * This method does all the job
+     * @param resultSet the info from Log Miner view
+     * @return number of processed DMLs from the given resultSet
+     */
+    int processResult(ResultSet resultSet) {
+        int dmlCounter = 0;
+        int commitCounter = 0;
+        int rollbackCounter = 0;
+        Duration cumulativeCommitTime = Duration.ZERO;
+        Duration cumulativeParseTime = Duration.ZERO;
+        Duration cumulativeOtherTime = Duration.ZERO;
+        Instant startTime = Instant.now();
+        while (true) {
```

Comment:
In this loop the redo log switching should also be checked. Not for every record because that would be to expensive but I do it every 10 seconds. Assuming the switching of all redo files takes more then 10 seconds to go around and reuse the active redo log.

Reply:
> In this loop the redo log switching should also be checked. Not for every record because that would be to expensive but I do it every 10 seconds. Assuming the switching of all redo files takes more then 10 seconds to go around and reuse the active redo log.

Reply:
Added as a follow-up under https://issues.redhat.com/browse/DBZ-2457.